src/execution/window_segment_tree.cpp

```diff
@@ -367,24 +367,248 @@ void WindowCustomAggregator::Evaluate(WindowAggregatorState &lstate, const DataC
 				frames[nframes++] = FrameBounds(cur_row, cur_row + 1);
 			}
 
-			//	WindowSegmentTreePart::RIGHT
+			//	WindowExcludePart::RIGHT
 			end = ends[i];
 			begin = (exclude_mode == WindowExcludeMode::CURRENT_ROW) ? (cur_row + 1) : peer_end[i];
 			begin = MinValue(begin, end);
 			non_empty += (begin < end);
 			frames[nframes++] = FrameBounds(begin, end);
 		}
 
+		operation(i, non_empty);
+	}
+}
+
+void WindowCustomAggregator::Evaluate(WindowAggregatorState &lstate, const DataChunk &bounds, Vector &result,
+                                      idx_t count, idx_t row_idx) const {
+	auto &lcstate = lstate.Cast<WindowCustomAggregatorState>();
+	auto &frames = lcstate.frames;
+	auto &rmask = FlatVector::Validity(result);
+	const_data_ptr_t gstate_p = nullptr;
+	if (gstate) {
+		auto &gcstate = gstate->Cast<WindowCustomAggregatorState>();
+		gstate_p = gcstate.state.data();
+	}
+
+	EvaluateSubFrames(bounds, exclude_mode, count, row_idx, frames, [&](idx_t i, bool non_empty) {
 		//	No data means NULL
 		if (!non_empty) {
 			rmask.SetInvalid(i);
-			continue;
+			return;
 		}
 
 		// Extract the range
 		AggregateInputData aggr_input_data(aggr.GetFunctionData(), lstate.allocator);
 		aggr.function.window(aggr_input_data, *partition_input, gstate_p, lcstate.state.data(), frames, result, i);
+	});
+}
+
+//===--------------------------------------------------------------------===//
+// WindowNaiveAggregator
+//===--------------------------------------------------------------------===//
+WindowNaiveAggregator::WindowNaiveAggregator(AggregateObject aggr, const LogicalType &result_type,
+                                             const WindowExcludeMode exclude_mode_p, idx_t partition_count)
+    : WindowAggregator(std::move(aggr), result_type, exclude_mode_p, partition_count) {
+}
+
+WindowNaiveAggregator::~WindowNaiveAggregator() {
+}
+
+class WindowNaiveState : public WindowAggregatorState {
+public:
+	struct HashRow {
+		explicit HashRow(WindowNaiveState &state) : state(state) {
+		}
+
+		size_t operator()(const idx_t &i) const {
+			return state.Hash(i);
+		}
+
+		WindowNaiveState &state;
+	};
+
+	struct EqualRow {
+		explicit EqualRow(WindowNaiveState &state) : state(state) {
+		}
+
+		bool operator()(const idx_t &lhs, const idx_t &rhs) const {
+			return state.KeyEqual(lhs, rhs);
+		}
+
+		WindowNaiveState &state;
+	};
+
+	using RowSet = std::unordered_set<idx_t, HashRow, EqualRow>;
+
+	explicit WindowNaiveState(const WindowNaiveAggregator &gstate);
+
+	void Evaluate(const DataChunk &bounds, Vector &result, idx_t count, idx_t row_idx);
+
+protected:
+	//! Flush the accumulated intermediate states into the result states
+	void FlushStates();
+
+	//! Hashes a value for the hash table
+	size_t Hash(idx_t rid);
```

Comment:
Nit: `size_t`

Reply:
Actually that is deliberate because the standard library containers expect that to be the return type.